redhat-appstudio/service-provider-integration-operator
# Add the ability to associate the secret generate by the binding with a potentially pre-existing service account.

### What does this PR do?
`$TITLE`

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-233

### How to test this PR?

Service account token:

1. create an empty service account:
   ```yaml
   apiVersion: v1
   kind: ServiceAccount
   metadata:
     name: test-sa
     namespace: default
   ```
2. create a binding with a service account definition:
    ```yaml
    ...
    spec:
       secret:
         type: kubernetes.io/service-account-token
         annotations:
           "kubernetes.io/service-account.name": "test-sa"
         linkedTo:
         - serviceAccount:
             reference:
               name: test-sa
    ```
3. upload the token data or go through the OAuth flow
4. check that the binding is injected and mentions the `serviceAccountNames` in its status.
5. check that the secret has the correct type and has the prescribed annotation.
6. check that the newly created service account mentions the secret in its `secrets` field.
7. check that the secret has the token in the `extra` field and `ca.crt`, `namespace` and `token` auto-filled by the Kubernetes control plane.
8. delete the binding
9. check that the secret has been deleted.
10. check that the service account is still there

(we do not support managed service account token secrets. These tokens have very limited utility in SPI, because all we can do is inject extra fields into them, while the main work is done by the Kubernetes infrastructure (injection of the actual service account token and certificate)).
 
Image pull secret:
1. Create the binding with:
   ```yaml
    ...
    spec:
       secret:
         type: kubernetes.io/dockerconfigjson
         linkedTo:
         - serviceAccount:
             as: imagePullSecret
             managed:
               generateName: test-sa-
    ```
3. upload the token data or go through the OAuth flow
6. check that the newly created service account mentions the secret in its `imagePullSecrets` field.
2. the service account should link the secret in its `imagePullSecrets` section.
8. delete the binding
9. check that both the secret and the service account have been deleted.

Other secret types:

It should be possible to link other secret types with the service account.

1. Repeat the steps for the image pull secrets but replace the secret type for `kubernetes.io/basic-auth` and remove the `as: imagePullSecret` definition.
5. check that the secret appears in the list of the secrets in the service account.
6. the secret should NOT have any new annotation (as opposed to the service account token secret)
7. no auto-filled stuff should be present in the secret.

Unmanaged:
1. Unmanaged means that the service account must already exist in the cluster, so let's just use the `default` SA.
1. Link some secret to the `default` SA manually (i.e. add a reference to the secret to the `secrets` field of the `default` SA).
1. Repeat the above steps for any secret type but replace the 
   ```yaml
   managed:
     generateName: ...
   ``` 
   with
   ```yaml
   reference:
      name: default
   ```
1. check that the binding secret is merely added to the list of the service acount secrets, not overwriting any pre-existing links.
1. The service account should stay there after the binding is deleted.

## Additional notes
There is an image with the latest code from this PR @ `quay.io/lkrejci/service-provider-integration-oauth:svpi-233`
